libtenzir/builtins/operators/from_to_2.cpp

```diff
@@ -582,9 +645,702 @@ class to_plugin2 final : public virtual operator_factory_plugin {
+          if (slash != std::string::npos) {
+            // The slash itself should be included.
+            return prefix->substr(0, slash + 1);
+          }
+        }
+      }
+      return "/";
+    });
+    query_files();
+  }
+
+  auto make_behavior() -> from_file_actor::behavior_type {
+    return {
+      [this](atom::get) -> caf::result<table_slice> {
+        return get();
+      },
+      [this](atom::put, table_slice slice) -> caf::result<void> {
+        return put(std::move(slice));
+      },
+      [this](diagnostic diag) {
+        dh_->emit(std::move(diag));
+      },
+      [this](uint64_t nested_operator_index, uuid nested_metrics_id,
+             type schema) {
+        return register_metrics(nested_operator_index, nested_metrics_id,
+                                std::move(schema));
+      },
+      [this](uint64_t nested_operator_index, uuid nested_metrics_id,
+             record metrics) {
+        return handle_metrics(nested_operator_index, nested_metrics_id,
+                              std::move(metrics));
+      },
+      [](const operator_metric& metrics) {
+        // Cannot forward operator metrics from nested pipelines.
+        TENZIR_UNUSED(metrics);
+      },
+    };
+  }
+
+private:
+  auto get() -> caf::result<table_slice> {
+    if (puts_.empty()) {
+      auto rp = self_->make_response_promise<table_slice>();
+      gets_.emplace_back(rp);
+      check_termination();
+      return rp;
+    }
+    auto slice = std::move(puts_.front().first);
+    puts_.front().second.deliver();
+    puts_.pop_front();
+    return slice;
+  }
+
+  auto put(table_slice slice) -> caf::result<void> {
+    if (gets_.empty()) {
+      auto rp = self_->make_response_promise<void>();
+      puts_.emplace_back(std::move(slice), rp);
+      return rp;
+    }
+    gets_.front().deliver(std::move(slice));
+    gets_.pop_front();
+    return {};
+  }
+
+  void query_files() {
+    add_actor_callback(
+      self_, fs_->GetFileInfoAsync(std::vector{root_path_}),
+      [this](arrow::Result<std::vector<arrow::fs::FileInfo>> infos) {
+        if (not infos.ok()) {
+          diagnostic::error("{}", infos.status().ToStringWithoutContextLines())
+            .primary(args_.url)
+            .emit(*dh_);
+          return;
+        }
+        TENZIR_ASSERT(infos->size() == 1);
+        auto root_info = std::move((*infos)[0]);
+        switch (root_info.type()) {
+          case arrow::fs::FileType::NotFound:
+            // TODO: Do we want to allow this? Maybe if `watch=true`?
+            diagnostic::error("`{}` does not exist", root_path_)
+              .primary(args_.url)
+              .emit(*dh_);
+            return;
+          case arrow::fs::FileType::Unknown:
+            diagnostic::error("`{}` is unknown", root_path_)
+              .primary(args_.url)
+              .emit(*dh_);
+            return;
+          case arrow::fs::FileType::File:
+            // TODO: Do we want to warn when the glob continues here?
```

Comment:
I'd say yes, unless we're watching.

Reply:
If the glob doesn't match and we are not watching, this is now an error (which is in line with how the other situations are handled, I think). When watching, no diagnostic is produced.